src/resmom/mom_comm.c

```diff
@@ -6156,14 +6524,332 @@ tm_request(int fd, int version)
 			(void)dis_flush(fd);
 			goto err;
 
+		case TM_SPAWN_MULTI:
+			/*
+			 ** Spawn a task on the requested node.
+			 **
+			 **	read (
+			 **		list_size	int;
+			 **		where 0		int;
+			 **		...
+			 **		where list_size-1	int;
+			 **		argc		int;
+			 **		arg 0		string;
+			 **		...
+			 **		arg argc-1	string;
+			 **		env 0		string;
+			 **		...
+			 **		env m		string;
+			 **	)
+			 */
+			DBPRT(("%s: SPAWN_MULTI %s\n",
+				__func__, jobid))
+
+			/* Get the number of nodes */
+			list_size = disrui(fd, &ret);
+			if (ret != DIS_SUCCESS)
+				goto err;
+
+			/* Get the list of where to send the executable */
+			node_list = (tm_node_id *)calloc(list_size, sizeof(tm_node_id));
+			assert(node_list);
+
+			for (i = 0; i < list_size; i++) {
+				node_list[i] = disrui(fd, &ret);
+				if (ret != DIS_SUCCESS) {
+					free(node_list);
+					goto err;
+				}
+			}
+			argc = disrui(fd, &ret);
+			if (ret != DIS_SUCCESS) {
+				free(node_list);
+				goto done;
+			}
+			argv = (char **)calloc(argc + 1, sizeof(char *));
+			assert(argv);
+			for (i = 0; i < argc; i++) {
+				argv[i] = disrst(fd, &ret);
+				if (ret != DIS_SUCCESS) {
+					argv[i] = NULL;
+					arrayfree(argv);
+					free(node_list);
+					goto done;
+				}
+				if(strlen(argv[i]) == 0)
+				  	found_empty_string = 1;  /* arguments contains empty string, Used if spawn on another MOM*/
+			}
+			argv[i] = NULL;
+				
+			numele = 3;
+			envp = (char **)calloc(numele, sizeof(char *));
+			assert(envp);
+			for (i = 0; ; i++) {
+				char	*env;
+
+				env = disrst(fd, &ret);
+				if (ret != DIS_SUCCESS && ret != DIS_EOD) {
+					arrayfree(argv);
+					envp[i] = NULL;
+					arrayfree(envp);
+					free(node_list);
+					goto done;
+				}
+				if (env == NULL)
+					break;
+				if (*env == '\0') {
+					free(env);
+					break;
+				}
+				/*
+				 **	Need to remember extra slot for NULL
+				 **	at the end.  Thanks to Pete Wyckoff
+				 **	for finding this.
+				 */
+				if (i == numele-1) {
+					numele *= 2;
+					envp = (char **)realloc(envp,
+						numele*sizeof(char *));
+					assert(envp);
+				}
+				envp[i] = env;
+			}
+			envp[i] = NULL;
+			ret = DIS_SUCCESS;
+
+			tempinfo = (struct job_multispawn_info *)calloc(1, sizeof(struct job_multispawn_info));
+			assert(tempinfo);
+			temptaskid = (tm_task_id *)calloc(list_size, sizeof(tm_task_id));
+			assert(temptaskid);
+			tempnid = (tm_node_id *)calloc(list_size, sizeof(tm_node_id));
+			assert(tempnid);
+
+			taskident++;
+			
+			tempinfo->ji_ident = taskident;
+			index = 0;
+			
+			/* open the tpp mcast channel */
+			if ((mtfd = tpp_mcast_open()) == -1) {
+				sprintf(log_buffer, "mcast open failed");
+				log_joberr(-1, __func__, log_buffer, jobid);
+				free(node_list);
```

Comment:
you could free them in the err block

Reply:
Since there are not many, I'd rather just take care of it here.  There is a history of each case statement taking care of freeing what it allocated.

Reply:
never mind, I decided it made sense to do it in err/done after all.